src/server/main_service.cc

```diff
@@ -724,14 +727,62 @@ void Service::Quit(CmdArgList args, ConnectionContext* cntx) {
 }
 
 void Service::Multi(CmdArgList args, ConnectionContext* cntx) {
-  if (cntx->conn_state.exec_state != ConnectionState::EXEC_INACTIVE) {
+  if (cntx->conn_state.exec_info.IsActive()) {
     return (*cntx)->SendError("MULTI calls can not be nested");
   }
-  cntx->conn_state.exec_state = ConnectionState::EXEC_COLLECT;
+  cntx->conn_state.exec_info.state = ConnectionState::ExecInfo::EXEC_COLLECT;
   // TODO: to protect against huge exec transactions.
   return (*cntx)->SendOk();
 }
 
+void Service::Watch(CmdArgList args, ConnectionContext* cntx) {
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  // Skip if EXEC will already fail due previous WATCH.
+  if (exec_info.watched_dirty.load(memory_order_relaxed)) {
+    return (*cntx)->SendOk();
+  }
+
+  atomic_uint32_t keys_existed = 0;
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    ArgSlice largs = t->ShardArgsInShard(shard->shard_id());
+    for (auto k : largs) {
+      shard->db_slice().RegisterWatchedKey(cntx->db_index(), k, &exec_info);
+    }
+
+    auto res = GenericFamily::OpExists(t->GetOpArgs(shard), largs);
+    keys_existed.fetch_add(res.value_or(0), memory_order_relaxed);
+    return OpStatus::OK;
+  };
+  cntx->transaction->ScheduleSingleHop(std::move(cb));
+
+  // Duplicate keys are stored to keep correct count.
+  exec_info.watched_existed += keys_existed.load(memory_order_relaxed);
+  for (size_t i = 1; i < args.size(); i++) {
+    exec_info.watched_keys.emplace_back(cntx->db_index(), ArgS(args, i));
+  }
+
+  return (*cntx)->SendOk();
+}
+
+// Unwatch all keys for a connection and unregister from DbSlices.
+// Used by UNWATCH, DICARD and EXEC.
+void UnwatchAllKeys(ConnectionContext* cntx) {
+  auto& exec_info = cntx->conn_state.exec_info;
+  if (!exec_info.watched_keys.empty()) {
+    auto cb = [&](EngineShard* shard) {
+      shard->db_slice().UnregisterConnectionWatches(&exec_info);
```

Comment:
does redis unwatch clear watched keys from all databases?

Reply:
Yes, this does as well